master/pkg/schemas/expconf/historical.go

```diff
@@ -0,0 +1,175 @@
+// current version of ExperimentConfig.  Fortunately, the parts of the system that need to read
+// configs that are this old are either 1) only concerned with the raw config for rendering it to
+// the user, or 2) only need very specific parts of the old config, so to reduce our own burden we
+// only expose very specific data from it.
+type HistoricalConfig struct {
+	checkpointStorage CheckpointStorageConfig
+	bindMounts        BindMountsConfig
+	envvars           EnvironmentVariablesMap
+}
+
+// CheckpointStorage returns a current CheckpointStorage from a HistoricalConfig.
+func (h HistoricalConfig) CheckpointStorage() CheckpointStorageConfig {
+	return h.checkpointStorage
+}
+
+// BindMounts returns a current BindMountsConfig from a HistoricalConfig.
+func (h HistoricalConfig) BindMounts() BindMountsConfig {
+	return h.bindMounts
+}
+
+// EnvironmentVariables returns a current EnvironmentVariables from a HistoricalConfig.
+func (h HistoricalConfig) EnvironmentVariables() EnvironmentVariablesMap {
+	return h.envvars
+}
+
+func getCheckpointStorage(raw map[string]interface{}) (CheckpointStorageConfig, error) {
+	cs := CheckpointStorageConfig{}
+
+	csOnly := raw["checkpoint_storage"]
+
+	csByts, err := json.Marshal(csOnly)
+	if err != nil {
+		return cs, errors.Wrap(err, "unable to remarshal config bytes as json")
+	}
+
+	// Read the checkpoint storage.
+	if err = schemas.SaneBytes(&cs, csByts); err != nil {
+		return cs, errors.Wrap(err, "historical checkpoint storage does not pass sanity checks")
+	}
+	if err = json.Unmarshal(csByts, &cs); err != nil {
+		return cs, errors.Wrap(err, "unable to unmarshal checkpoint storage bytes")
+	}
+
+	// Fill defaults (should be a no-op).
+	cs = schemas.WithDefaults(cs).(CheckpointStorageConfig)
+
+	// Validate fully before passing anything out.
+	if err = schemas.IsComplete(cs); err != nil {
+		return cs, errors.Wrap(err, "historical checkpoint storage is incomplete")
+	}
+
+	return cs, nil
+}
+
+func getBindMounts(raw map[string]interface{}) (BindMountsConfig, error) {
+	bm := BindMountsConfig{}
+
+	bmOnly := raw["bind_mounts"]
+	if bmOnly == nil {
+		// Empty bind_mounts.
+		return bm, nil
+	}
+
+	bmByts, err := json.Marshal(bmOnly)
+	if err != nil {
+		return bm, errors.Wrap(err, "unable to remarshal bind mounts as json")
+	}
+	if err = schemas.SaneBytes(&bm, bmByts); err != nil {
+		return bm, errors.Wrap(err, "historical bind mounts does not pass sanity checks")
+	}
+	if err = json.Unmarshal(bmByts, &bm); err != nil {
+		return bm, errors.Wrap(err, "unable to unmarshal bind mounts bytes")
+	}
+	bm = schemas.WithDefaults(bm).(BindMountsConfig)
+	if err = schemas.IsComplete(bm); err != nil {
+		return bm, errors.Wrap(err, "historical bind mounts is incomplete")
+	}
+	return bm, nil
+}
+
+func getEnvironmentVariables(raw map[string]interface{}) (EnvironmentVariablesMap, error) {
+	ev := EnvironmentVariablesMap{}
```

Comment:
how far are we from schemas being magical enough to just unmarshal the bytes into `HistoricalConfig` and then 
checking sane bytes/iscomplete/defaulting as usually after deser? i guess it just would require adding a historical config schema?

Reply:
I don't think it's a good idea to try to write a json schema HistoricalConfig.  I think in the future, we can check the version field and unmarshal that version, but the v0 handling is always going to be tricky.